codice/ddf
# DDF-3987 Build offline gazetteer suggester index whenever gazetteer metacards are created, updated, or deleted

#### What does this PR do?
- Changes when the Solr gazetteer suggester index is built. Currently the index is built on every Solr restart and `catalog ` core reload, regardless of whether the gazetteer entries have changed. This PR sets `buildOnStartup=false` and adds a new `PostIngestPlugin` that builds the suggester index when new gazetteer metacards are created, updated, or deleted.
- Adds a `gazetteer:build-suggester-index` command for manually issuing requests to build the suggester index (useful if the request submitted after the gazetteer update fails).
- Changes `storeDir` to `indexPath` in the suggest component in `solrconfig.xml`. `BlendedInfixLookupFactory` uses `indexPath` to store the suggester index. This change prevents Solr from printing a bogus failure message in its log and ensures the suggester index will be stored in the `place_suggest_index` directory.

#### Who is reviewing it? 
@adimka 
@bdeining 
@troymohl 

#### Select relevant component teams: 
@codice/solr 

#### Ask 2 committers to review/merge the PR and tag them here.
@millerw8 
@pklinef

#### How should this be tested?
- Start DDF and run through the installer. Restart DDF, open Intrigue, and create an advanced search. Select the `anyGeo` field then click the `Keyword` box. Click `Enter a region, country, or city` and type a few characters. Verify that matching suggestions are provided below. Also try searching for "Omaha" and verify that there are no results (this is relevant to step 3).
- Check Solr's log file `<DDF_INSTALL_LOCATION>/solr/server/logs/solr.log` and verify that `SolrSuggester.build(suggestPlace)` only appears once. This is printed every time the suggester index is built.
- Add more entries to the gazetteer using the `gazetteer:update` command and a dump file obtained from [GeoNames](http://download.geonames.org/export/dump/), such as `cities15000.zip`. Wait one minute (how long it takes for the suggester index to be built after ingesting new gazetteer entries), then go back to Intrigue and try searching for `Omaha` in the box you typed in in step 1 (Omaha is not included in the default geonames zip included with DDF but is included in `cities15000.zip`). Verify that you get at least one result back, which means the suggester index was updated after you added gazetteer entries.

#### What are the relevant tickets?
[DDF-3987](https://codice.atlassian.net/browse/DDF-3987)

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
